greenplumpython/func.py

```diff
@@ -170,292 +158,325 @@ def serialize(self) -> str:
                     s = to_pg_const(self._args[i])  # type: ignore
                 args_string_list.append(s)
             args_string = ",".join(args_string_list)
-        return f"{self._func_name}({args_string})"
+        return f"{self._func.qualified_name}({args_string})"
 
     def __call__(self, group_by: Optional[TableRowGroup] = None, table: Optional[Table] = None):
         return ArrayFunctionExpr(
-            self._func_name,
+            self._func,
             self._args,
             group_by=group_by,
-            as_name=self._as_name,
             table=table,
             db=self._db,
-            is_return_comp=self.is_return_comp,  # type: ignore
+            returning_composite=self._returning_composite,
         )
 
 
-def function(name: str) -> Callable[..., FunctionExpr]:
-    """
-    A wrap in order to call function
+# The parent class for all database functions.
+# It is not a Callable by design to prevent misuse.
+class _AbstractFunction:
+    def __init__(
+        self,
+        wrapped_func: Optional[Callable[..., Any]],
+        name: Optional[str],
+        schema: Optional[str],
+    ) -> None:
+        NAMEDATALEN = 64  # See definition in PostgreSQL
```

Comment:
NAMEDATALEN move to const not in this Class scope maybe better?

Reply:
I think here is better, limiting the scope first and relaxing it when needed is always better IMO